common/flatpak-utils.c

```diff
@@ -6402,12 +6404,169 @@ flatpak_progress_new (FlatpakProgressCallback progress,
+  GVariant *variant = ostree_async_progress_get_variant (source, key); \
+  if (variant)                                                         \
+    ostree_async_progress_set_variant (dest, key, variant);            \
+} G_STMT_END
+
+  COPY_OSTREE_STATE ("outstanding-metadata-fetches");
+  COPY_OSTREE_STATE ("outstanding-writes");
+  COPY_OSTREE_STATE ("scanning");
+  COPY_OSTREE_STATE ("scanned-metadata");
+  COPY_OSTREE_STATE ("fetched-delta-parts");
+  COPY_OSTREE_STATE ("total-delta-parts");
+  COPY_OSTREE_STATE ("fetched-delta-fallbacks");
+  COPY_OSTREE_STATE ("total-delta-fallbacks");
+  COPY_OSTREE_STATE ("fetched-delta-part-size");
+  COPY_OSTREE_STATE ("total-delta-part-size");
+  COPY_OSTREE_STATE ("bytes-transferred");
+  COPY_OSTREE_STATE ("fetched");
+  COPY_OSTREE_STATE ("metadata-fetched");
+  COPY_OSTREE_STATE ("requested");
+  COPY_OSTREE_STATE ("start-time");
+  COPY_OSTREE_STATE ("status");
+  COPY_OSTREE_STATE ("total-extra-data-bytes");
+  COPY_OSTREE_STATE ("transferred-extra-data-bytes");
+  COPY_OSTREE_STATE ("downloading-extra-data");
+
+  /* Copy these last, since progress_cb() considers the object "ready" when
+   * they are present */
+  COPY_OSTREE_STATE ("outstanding-extra-data");
+  COPY_OSTREE_STATE ("outstanding-fetches");
+
+#undef COPY_OSTREE_STATE
+}
+
+#endif  /* libostree < 2019.6 */
+
+static void
+handle_chained_progress (OstreeAsyncProgress *chained_progress,
+                         gpointer             user_data)
+{
+  OstreeAsyncProgress *original_progress = (OstreeAsyncProgress *) user_data;
+
+  /* Sync the chained progress's state back to the original instance, to take
+   * into account any updates received while a different GMainContext was
+   * active */
+#if OSTREE_CHECK_VERSION(2019, 6)
+  ostree_async_progress_copy_state (chained_progress, original_progress);
+#else
+  copy_state_fallback (chained_progress, original_progress);
+#endif
+
+  OstreeAsyncProgress *chained_from =
+    OSTREE_ASYNC_PROGRESS (g_object_get_data (G_OBJECT (original_progress), "chained_from"));
+  FlatpakProgressCallback chained_callback =
+    g_object_get_data (G_OBJECT (original_progress), "callback");
+
+  if (chained_from != NULL)
+    {
+      /* It's possible we chained to an already-chained progress object. */
+      handle_chained_progress (original_progress, chained_from);
+    }
+  else if (chained_callback != NULL)
+    {
+      /* The normal case; we chained to a progress object created by
+       * flatpak_progress_new(). */
+      gpointer original_data = g_object_get_data (G_OBJECT (original_progress), "callback_data");
+      progress_cb (original_progress, original_data);
+    }
+  else
+    {
+      /* Do nothing. It's possible we chained to a progress object without the
+       * GObject data, that was not created by flatpak_progress_new().
+       * Unfortunately it doesn't seem possible to call the callback in this
+       * case. */
```

Comment:
Maybe this should be a `g_assert_not_reached()`?

Reply:
Unfortunately, it is reached, that's why this comment is here :-/

For example the progress object created with `ostree_async_progress_new()` at the top of `_flatpak_dir_fetch_remote_state_metadata_branch()` is passed to `flatpak_dir_pull()` where it is chained and from there to `repo_pull()` where it is chained again.

This particular instance of OstreeAsyncProgress is probably not even necessary, since a callback is never connected to it. Another option would be to add the assert, and find all such instances and just pass `NULL` progress since the callback is never connected in that case, but that seems more risky since I don't know whether I'd be testing all the possible code paths.